### PR TITLE
GSoC 2018: Add Tours

### PR DESCRIPTION
**Pull request by JoshuaBonn1**
See https://musescore.org/en/node/274365 for a guide and more explanation
Blog Posts:
Week 6: https://musescore.org/en/user/1190941/blog/2018/06/25/gsoc-2018-beginner-mode-and-tutorial-creation-week-6
Week 7: https://musescore.org/en/user/1190941/blog/2018/07/02/gsoc-2018-beginner-mode-and-tutorial-creation-week-7
Week 8: https://musescore.org/en/user/1190941/blog/2018/07/09/gsoc-2018-beginner-mode-and-tutorial-creation-week-8
Week 9: https://musescore.org/en/user/1190941/blog/2018/07/16/gsoc-2018-beginner-mode-and-tutorial-creation-week-9

### CODE REVIEW

File: mscore/preferences.cpp [hunk 1]
```diff
@@ -92,6 +92,7 @@ void Preferences::init(bool storeInMemoryOnly)
             {PREF_APP_STARTUP_FIRSTSTART,                          new BoolPreference(true)},
             {PREF_APP_STARTUP_SESSIONSTART,                        new EnumPreference(QVariant::fromValue(SessionStart::SCORE), false)},
             {PREF_APP_STARTUP_STARTSCORE,                          new StringPreference(":/data/My_First_Score.mscz", false)},
+            {PREF_UI_APP_STARTUP_SHOWTOURS,                               new BoolPreference(true)},
```
Comment (MarcSabatella) [inline mscore/preferences.cpp]: indent off?

File: mscore/startupWizard.cpp [hunk 2]
```diff
@@ -156,6 +156,24 @@ StartupWizardPage3::StartupWizardPage3(QWidget* parent)
       setLayout(layout);
       }
 
+StartupWizardPage4::StartupWizardPage4(QWidget* parent)
+      : QWizardPage(parent)
+      {
+      setTitle(tr("Tours"));
+      QLabel* label = new QLabel(tr("Do you want to see helping tours?"), this);
```
Comment (MarcSabatella) [inline mscore/startupWizard.cpp]: I suppose we need to settle on terminology and be consistent, at least for user-visible strings.  The words "help", "tour", "tip", and "tutorial" all seem possibly relevant here.  Probably there are other terms that could be used too.  I don't really have a preference myself, but "helping tours" seems awkward.

File: mscore/tourhandler.cpp [hunk 3]
```diff
@@ -0,0 +1,460 @@
+#include "tourhandler.h"
```
Comment (MarcSabatella) [inline mscore/tourhandler.cpp]: I really like how few changes are required elsewhere in the code, so most of the work is done here in this tourhandler.cpp

File: mscore/startupWizard.cpp (hunk 2 above)
Comment (MarcSabatella) [reply mscore/startupWizard.cpp]: I think MuseScore isn't the only application to do this sort of thing, let's be consistent with established terminology.

Comment (JoshuaBonn1) [reply mscore/startupWizard.cpp]: Many of the examples I am seeing will not ask the users if they want to see the tours. Rather, they will show a welcome tour and give the user the option to decline all tours then.

**Comment by JoshuaBonn1** [comment]
As soon as this passes the tests, it is ready for further review and merging.

File: mscore/preferences.cpp [hunk 4]
```diff
@@ -92,6 +92,7 @@ void Preferences::init(bool storeInMemoryOnly)
             {PREF_APP_STARTUP_FIRSTSTART,                          new BoolPreference(true)},
             {PREF_APP_STARTUP_SESSIONSTART,                        new EnumPreference(QVariant::fromValue(SessionStart::SCORE), false)},
             {PREF_APP_STARTUP_STARTSCORE,                          new StringPreference(":/data/My_First_Score.mscz", false)},
+            {PREF_UI_APP_STARTUP_SHOWTOURS,                        new BoolPreference(true)},
```
Comment (Marr11317) [inline mscore/preferences.cpp]: Could tours be disabled by default in development versions?
We're always hitting the "revert to factory settings" function, so always passing through the tours again and again could be annoying. (plus, I don't believe MuseScore developers really need those tours)
Something like:
```cpp
#if THIS_IS_A_DEVELOPMENT_VERSION
{PREF_UI_APP_STARTUP_SHOWTOURS, new BoolPreference(false)},
#else
{PREF_UI_APP_STARTUP_SHOWTOURS, new BoolPreference(true)},
#endif
```

File: mscore/tourhandler.h [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+namespace Ms {
+
+//---------------------------------------------------------
+//   TourMessage
+//---------------------------------------------------------
+
+struct TourMessage {
+      QString message;
+      QList<QString> widgetNames;
+      void init(QString m, QList<QString> w) { message = m; widgetNames = w; }
```
Comment (Otto-AA) [inline mscore/tourhandler.h]: _Disclaimer: I am definitely no pro in C++, so feel free to disagree._

From my point of view it seems more concise to use a constructor and initialization list here:
```cpp
TourMessage(QString m, QList<QString> w) : message(m), widgetNames(w) {}
```

File: mscore/preferences.cpp (hunk 4 above)
Comment (JoshuaBonn1) [reply mscore/preferences.cpp]: You can disable tours in the startup wizard and on each tour. No more than 2 extra clicks.

Comment (MarcSabatella) [reply mscore/preferences.cpp]: FWIW, one could express the same concern about the other startup wizard for selecting config options etc.  It's a valid concern, I guess, but also not a big deal - I rather don't like having development buildsdeliberately behave different form release builds.  On the other hand, maybe a command line option to disable the tours could be worthwhile?

Comment (Marr11317) [reply mscore/preferences.cpp]: A command line option to disable the tours? It seems a little bit too specific to me. Maybe, we'll just get over it, and disable the tours after each 'revert to factory settings'.

Comment (MarcSabatella) [reply mscore/preferences.cpp]: Well, we do have command line options to disable a few other things that kind of sort of seem similar.  But mostly, I think it's just something we should just get over, as you say :-)

File: mscore/tourhandler.cpp [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+      {
+      QStringList nameFilters;
+      nameFilters << "*.tour";
+      QString path = mscoreGlobalShare + "tours";
+
+      QDir dir(path);
+      QStringList pl = dir.entryList(nameFilters, QDir::Files, QDir::Name);
+
+      for (const QString& entry : pl) {
+            QFile* tourFile = new QFile(path + "/" + entry);
```
Comment (MarcSabatella) [inline mscore/tourhandler.cpp]: Recently some issues were reported regarding how temporary filenames were being generated for the new extension facility - something about how the path separators were used was breaking things on Linux.  I didn't totally follow that, but please verify this is the recommended way to generate pathnames.

**Comment by MarcSabatella** [comment]
See my comment about assembling pathnames, also worth considering if the filenames are what we want (eg, "completedTours.list"), although of course we can always change them later.

Aside from that, all looks good to me.  Usually we recommend squashing commits, and that does apply here, but I'll actually make another related request - see if you can remove the additional sample tours from the main commit, and make a separate commit just for those.  Keep anything that *needs* to be there in order for the code to work (the welcome tour, I guess, since it is loaded explicitly) but it might be nice to have the others in a separate commit.

File: mscore/tourhandler.cpp (hunk 6 above)
Comment (JoshuaBonn1) [reply mscore/tourhandler.cpp]: The only other way I've been seeing (more often than the one I used here) is to use QDirIterator. I'll use that and also use QDirIterator::Subdirectories so they can be further sorted.

Comment (anatoly-os) [reply mscore/tourhandler.cpp]: No, the only way we separate directories in paths is "/" symbol. https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules#Pathnames

Comment (MarcSabatella) [reply mscore/tourhandler.cpp]: OK, thanks for clarifying.  I knew I had seen something about this, didn't realize the bottom line was that we changed *from* the QDir method to just using "/"

**Comment by Jojo-Schmitz** [comment]
Rebase needed

**Comment by MarcSabatella** [comment]
@JoshuaBonn1 in your blog you mentioned having build problems - is the rebase what that is about?  Let's make sure we get this taken care of ASAP, since we're into the home stretch on GSoC.  You will want everything building and as clean as possible when you submit your final code.  I'll be around and available on IRC most of today and tomorrow.

**Comment by JoshuaBonn1** [comment]
@MarcSabatella I realized it wasn't building because some of the components were on the wrong version (JACK, etc.). I've been keeping everything clean and ready to merge (to the best of my abilities), waiting for it to be selected.

**Comment by MarcSabatella** [comment]
Great!  As far as I am concerned, this is ready to be merged.

**Comment by anatoly-os** [comment]
I've tested the tour a bit. It is awesome! It is definitely a good start of creating the best FTUX in MuseScore.
Btw, I found one bug: the tour widgets interfere with StartCenter widget, so that when changing tour screens, startCenter blinks in background.
And one more suggestion/question: why did you place "Next" button to the left, but "Back" to the right? It looks counterintuitive. Also, where can I find the whole list of widgets which are covered by the tour?

**Comment by JoshuaBonn1** [comment]
@anatoly-os Strange bug. If I can find time while schooling, I'll check it out. As for the placement of the Next/Back button, see this video: https://www.youtube.com/watch?v=jWYEvNTjlus&feature=youtu.be . Let me know if this matches what you see or not. Lastly, there is no current way to view every widget covered by the tour. The only similar thing you can do is look at each message and see all the widgets attached to the specific message.

**Comment by anatoly-os** [comment]
This is how it looks on Windows:
image

**Comment by anatoly-os** [comment]
@JoshuaBonn1 are you around to fix the bug with buttons placement? It should be easy due to special settings in Qt which define buttons position depending on the system. Something which is done in https://github.com/musescore/MuseScore/pull/3867/files may help.

**Comment by MarcSabatella** [comment]
@JoshuaBonn1 Let me know if you think you'll be able to look at this further, and if not, if there is you have insight you can share with me.  I'd like to see this merged.  Then hopefully I can others can come up with a plan to develop some content!

**Comment by JoshuaBonn1** [comment]
@MarcSabatella Sorry, I've been quite busy with school work. I had to uninstall my development stuff for MuseScore due to space restraints. And now, my computer has stopped working entirely :/ It looks like anatoly-os has the correct idea. I might be able to find the exact location where that needs to be put, but I still can't build it myself.

**Comment by anatoly-os** [comment]
One more thing. Now tour widget appears along with Start Center. I suggest showing first tour message right after closing Start Center.

File: share/tours/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,28 @@
+#=============================================================================
+#  MusE
```
Comment (anatoly-os) [inline share/tours/CMakeLists.txt]: This is definitely not the actual copyrights text.

**Comment by anatoly-os** [comment]
Do we translate tour text somehow? I don't see the translation code in this PR.

**Comment by Jojo-Schmitz** [comment]
Texts in the code (`.cpp` and `.ui`) are translatable, but those `.tour` files would need to get translated resp. to be made translatable

**Comment by anatoly-os** [comment]
I had a chance to look closely at the code. The following improvements should be done before merging (and one after):
1. Tours should take Start Center running into account. If start center is run, tours should connect to the `closed()` signal of the start center and appear after Start Center has been closed. It won't work if users disabled Start Center on startup. In this case, you need to check the bool variable checking whether Start Center is opened on startup or not and start tours directly if running Start Center on startup is disabled.
2. Buttons position. Buttons work correctly in terms the way you programmed them. The problem is that Yes/No button roles are positioned differently on Mac/Linux and Windows. But next/previous buttons position should always be the same. So, you don't need to specify Yes/No roles for the buttons.
3. (after merge) You need to find the way to translate strings from the tours. Should we use Qt engine+transifex for that? I think so.

**Comment by anatoly-os** [comment]
See #4054 addressing 1. and 2.

**Comment by anatoly-os** [comment]
@JoshuaBonn1 @MarcSabatella congratulations! Your project has been merged to MuseScore, so everyone will see your input to the project :)

**Comment by MarcSabatella** [comment]
Fantastic!  Now I look forward to working on a strategy to actually develop the tours :-)

**Pull request closed without merging**